Sources/_Datadog_Private/include/Swizzler.h

```diff
@@ -0,0 +1,30 @@
+//
+//  Swizzler.h
+//  Datadog
+//
+//  Created by Mert Buran on 11/05/2020.
+//  Copyright © 2020 Datadog. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface Swizzler : NSObject
+
++ (BOOL)    swizzle:(NSURLSession *)session
+ requestInterceptor:(NSURLRequest *(^)(NSURLRequest *))requestInterceptor
+       taskObserver:(void(^)(NSURLSessionTask *))taskObserver
+enforceDynamicClassCreation:(BOOL)enforceCreateNewClass
+              error:(NSError **)error;
```

Comment:
This violates the open-closed principle and won't scale for other features (like RUM auto-instrumentation for `UIViewController`), as swizzling new method requires:
* ~100 lines of very complex code to be added to `Swizzler`;
* even more code for tests;
* testing each swizzling individually.

Current implementation of `Swizzler` is tightly coupled with `URLSession`, but what we need is a domain-agnostic swizzling.


#### Domain-agnostic swizzling

What we need is to:
* _inherit part of a "source object" interface with methods coming from a "template object"_.

A specialisation of this requirement would be what we need toauto-instrumentation the `URLSession`:
* _inherit part of the `URLSession` with methods coming from the `TemplateURLSession`_.

 This sort of API would be properly opened for extensions and closed for modification:
```objc
/// Swizzles `sourceObject` to call corresponding method in `templateObject`
///  before executing the base method in `sourceObject`.
+(BOOL)  inherit:(NSObject *)sourceObject 
byTemplateObject:(NSObject *)templateObject
           error:(NSError **)error;
```

And the `URLSession` use case will be reduced to this:
```swift
let session = URLSession.shared
let sessionTemplate = URLSessionTemplate(requestInterceptor: RequestInterceptor())

try Swizzler.inherit(session, byTemplateObject: sessionTemplate)
```

With this approach, the unit tests will be reduced to:
```swift
class SourceObject: NSObject {
   var someMethodWasCalled = false

   func someMethod() { someMethodWasCalled = true }
}

class TemplateObject: NSObject {
   var someMethodWasCalled = false

   func someMethod() { someMethodWasCalled = true }
}

func testWhenCallingMethodOnASourceObject_theCorrespondingMethodIsCalledOnTemplateObejct() {
   // Given
   let sourceObject = SourceObject()
   let templateObject = TemplateObject()

   try Swizzler.inherit(sourceObject, byTemplateObject: templateObject)

   // When
   sourceObject.method1()

   // Then
   XCTAssertTrue(sourceObject.someMethodWasCalled)
   XCTAssertTrue(templateObject.someMethodWasCalled)
}
```
---
Future use cases in RUM will leverage the code reuse:
```swift
try Swizzler.inherit(viewController, byTemplateObject: viewControllerTemplate)
try Swizzler.inherit(navigationController, byTemplateObject: navigationController)
try Swizzler.inherit(button, byTemplateObject: uiControlTemplate)
```

Reply:
### `NSURLSession` Coupling
`Swizzler.h` might be coupled to `URLSession` but `Swizzler.m` is not really. The only part which is `URLSession`-specific is where we add `injected_` methods (ie: `requestInterceptor/taskObserver`), the rest doesn't have `URLSession` references.

`Swizzler.h` has `URLSession` references for the clarity; it would still work if we renamed it to `swizzle:(id)object...`

### Testing

Since we don't have type safety, we need to test every swizzled class/method one by one; there is no shortcut.
If the `Swizzler` could take any kind of object we would need to test every kind of objects; as we only need to swizzle `URLSession` it only takes `session` and we only test `session`s

As a side-note: `TemplateURLSession` class is written only for providing some sort of type safety at compile time, we could write `Swizzler` without it either but then we would need to rely on free functions with probably less type safety.

### Reusability

It's hard to foresee the other use cases for swizzling. `Swizzler` does `isa swizzling`, we will probably not need it for `UIViewController` and go with `method swizzling` as we don't change the execution path unlike in `URLSession`

#### If we swizzle methods of `UIViewController`

We will probably add a new method for method swizzling to `Swizzler` but won't share much between method and isa swizzling.

#### If we `isa` swizzle `UIViewController`

We will need to modify `injected_` methods (as we won't intercept requests in `ViewController`) and keep using the rest of `Swizzler` code

Either way we will need to test every swizzled method, subclass, etc.
Having a generic swizzler won't change that unfortunately.

## Wrap up

At first, injecting blocks into a dynamic class to customize behavior sounded like a good idea to me but now I find it confusing. We can implement them in `TemplateClass` at compile-time and that would be less confusing.
When I do that, `Swizzler` will be working as you described above (since that was the only difference)

However, I prefer keeping `TemplateClass` as a class constant, making it a variable would increase complexity by a lot.
Maybe we can subclass `Swizzler` and override `TemplateClass` such as

```objc
class URLSessionSwizzler: Swizzler {
  override class var templateClass: AnyClass { return URLSession.self }
}
class FooSwizzler: Swizzler {
  override class var templateClass: AnyClass { return Foo.self }
}
```

I'm making the change now 🛠

Reply:
> Swizzler.h might be coupled to URLSession but Swizzler.m is not really. The only part which is URLSession-specific is where we add injected_ methods (ie: requestInterceptor/taskObserver), the rest doesn't have URLSession references.

> Swizzler.h has URLSession references for the clarity; it would still work if we renamed it to swizzle:(id)object...

Reply:
> Swizzler.h might be coupled to URLSession but Swizzler.m is not really. The only part which is URLSession-specific is where we add injected_ methods (ie: requestInterceptor/taskObserver), the rest doesn't have URLSession references.

> Swizzler.h has URLSession references for the clarity; it would still work if we renamed it to swizzle:(id)object...

So it sounds like decoupling `injected_` methods from `Swizzler.m` and moving `URLSession` references away from `Swizzler.h` will make it generic ✨. I think the solution with subclassing the `Swizzler` may make this thing right 👌.

> Since we don't have type safety, we need to test every swizzled class/method one by one; there is no shortcut.
If the Swizzler could take any kind of object we would need to test every kind of objects; as we only need to swizzle URLSession it only takes session and we only test sessions

This is totally fine 👌. Let's just create easy solution for testing swizzled things.

The first step IMO is to separate "`Swizzler` tests", from "`URLSession` swizzling tests". Now everything is mixed together in `SwizzlerTests`. Once we have separate test suites, we may look into it and see if some generic helpers can be introduced to make next swizzlings in our codebase trivial. How about 3 levels of tests:
* 1/ `SwizzlerTests` - **tests swizzling** of basic `NSObject`;
* 2/ `URLSessionSwizzlingTests` - **tests the type safety** of swizzled methods - here generic helpers can be a real booster;
* 3/ `URLSessionInterceptorTests` - **tests the logic of intercepting** `URLSession` calls and adding trace information.

> Either way we will need to test every swizzled method, subclass, etc.
Having a generic swizzler won't change that unfortunately.

With the above testing model, new use cases will require only 2nd layer for type safety and 3rd layer for logic tests.

> **If we swizzle methods of UIViewController**
> We will probably add a new method for method swizzling to Swizzler but won't share much between method and isa swizzling.

You're right 👍. Method swizzling is a separate thing that may be added to `Swizzler`, or we may want to separate `MethodSwizzler` from `ISASwilzzler`.

> **If we isa swizzle UIViewController**
> We will need to modify injected_ methods (as we won't intercept requests in ViewController) and keep using the rest of Swizzler code

Yes, so the `Swizzler` is not open for extensions and not closed for modifications. That's why let's generalize it 🚀.

> At first, injecting blocks into a dynamic class to customize behavior sounded like a good idea to me but now I find it confusing. We can implement them in TemplateClass at compile-time and that would be less confusing.
When I do that, Swizzler will be working as you described above (since that was the only difference)

IMO it's better to inject an object (`RequestInterceptor` or sth), instead of separate closures for different things. Single object will be much easier to work with in tests, and way easier to add thread safety if necessary.

> However, I prefer keeping TemplateClass as a class constant, making it a variable would increase complexity by a lot.

Yes, totally, I really like the template class 🙌 . Just keep in mind that we'd probably need more template classes. Even in current use case, I think we need `URLSessionTemplate` and `URLSessionTaskTemplate`, so this way we can reuse `Swizzler` to swizzle `task.resume()`. Although, I'm not yet sure how the completion may be captured on `task`.